torch/distributed/_mesh_layout.py

```diff
@@ -59,13 +75,135 @@ def __getitem__(self, i: int) -> "_MeshLayout":
         return _MeshLayout(layout.shape, layout.stride)
 
     def coalesce(self) -> "_MeshLayout":
+        """
+        A layout is represented by (sizes):(strides), e.g. (3,2):(4,2).
+        Two consecutive dimensions can be "merged" into one if their
+        strides are contiguous/multiplicative (i.e., the inner stride * inner size
+        equals the next stride), we perform this kind of merge inside coalesce.
+
+        Example 1 (simple): (3,2):(2,1)
+        - inner dimension: has stride=1, size=2
+        - outer dimension: stride = inner_stride * inner_size = 2
+        → coalesced = (6:1)    # acts like a flat 1D array of length 6
+
+        Example 2 (non-coalescible): (3,2):(4,1)
+        - inner dimension: stride=1, size=2 → 2*1 = 2
+        - outer dimension: stride=4, mismatch (≠ 2)
+        → cannot merge; result stays (3,2):(4,1)
+        """
         layout = coalesce(self)
         return _MeshLayout(layout.shape, layout.stride)
 
     def composition(self, layout: "_MeshLayout") -> "_MeshLayout":
+        """
+        Perform a by-dimension composition between this layout (self) and another layout (layout).
+
+        Mental model about how to understand the composition logic:
+        - Think of each dimension in this layout as a "slot" that can itself be
+          refined by another layout.
```

Comment:
My problem with this intuition is that you could say the same thing about concatenation. I prefer the intuition where composition is about "picking" elements out of the lhs's domain. You only ever generate indices that the lhs function could generate but the rhs function decides which particular one you pick for any given input